node-graph/gcore/src/graphic_element.rs

```diff
@@ -0,0 +1,179 @@
+use core::ops::{Deref, DerefMut};
+
+use crate::Node;
+use dyn_any::{DynAny, StaticType};
+use glam::IVec2;
+use node_macro::node_fn;
+
+use crate::{
+	raster::{BlendMode, ImageFrame},
+	vector::VectorData,
+	Color,
+};
+
+/// A list of [`GraphicElement`]s
+#[derive(Clone, Debug, Hash, PartialEq, DynAny, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct GraphicGroup(Vec<GraphicElement>);
+
+/// Internal data for a [`GraphicElement`]. Can be [`VectorData`], [`ImageFrame`], text or a [`GraphicGroup`]
+#[derive(Clone, Debug, Hash, PartialEq, DynAny)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum GraphicElementData {
+	VectorShape(Box<VectorData>),
+	ImageFrame(ImageFrame<Color>),
+	Text(String),
+	GraphicGroup(GraphicGroup),
+}
+/// A named [`GraphicElementData`] with a blend mode, opacity, visibility, locked and collapsed state.
+#[derive(Clone, Debug, PartialEq, DynAny)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct GraphicElement {
+	pub name: String,
+	pub blend_mode: BlendMode,
+	pub opacity: f32,
+	pub visible: bool,
+	pub locked: bool,
+	pub collapsed: bool,
+	pub graphic_element_data: GraphicElementData,
+}
+
+/// Some [`ArtboardData`] with some optional clipping bounds and a label, that can be exported.
+/// Similar to an inkscpae page: https://media.inkscape.org/media/doc/release_notes/1.2/Inkscape_1.2.html#Page_tool
+#[derive(Clone, Debug, Hash, PartialEq, DynAny)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Artboard {
+	pub artboard_data: ArtboardData,
+	pub label: String,
+	pub bounds: Option<[IVec2; 2]>,
+}
+
+/// A list of [`Artboard`]s
+#[derive(Clone, Debug, Hash, PartialEq, DynAny, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ArtboardGroup(Vec<Artboard>);
+
+/// Either a [`GraphicGroup`] or a nested [`ArtboardGroup`].
+#[derive(Clone, Debug, Hash, PartialEq, DynAny)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ArtboardData {
+	GraphicGroup(GraphicGroup),
+	ArtboardGroup(ArtboardGroup),
+}
+
+pub struct ConstructLayerNode<Name, BlendMode, Opacity, Visible, Locked, Collapsed, Stack> {
+	name: Name,
+	blend_mode: BlendMode,
+	opacity: Opacity,
+	visible: Visible,
+	locked: Locked,
+	collapsed: Collapsed,
+	stack: Stack,
+}
+
+#[node_fn(ConstructLayerNode)]
+fn construct_layer<Data: Into<GraphicElementData>>(
+	graphic_element_data: Data,
+	name: String,
+	blend_mode: BlendMode,
+	opacity: f32,
+	visible: bool,
+	locked: bool,
+	collapsed: bool,
+	mut stack: GraphicGroup,
+) -> GraphicGroup {
+	stack.push(GraphicElement {
+		name,
+		blend_mode,
+		opacity,
```

Comment:
Can you multiply/divide opacity by 100 on this line, instead of having to make it part of the number input widget? I'm not sure I really like requiring the number input builder function to take the `1.` input from every caller when it's rare to require a multiplier.